Devel-PPPort: fix STMT_START and STMT_END to not warn on clang

haarg opened the pull request:
Since 7169efc77525df70484a824bff4ceebd1fafc760, perl's core STMT_START and STMT_END macros no longer try to use brace groups, due to the warnings they can generate and their very limited usefulness.

That commit also changed Devel::PPPort to remove the use of brace groups in STMT_START/STMT_END. That led to errors in older perls, so it was partly reverted in e08ee3cb66f362c4901846a46014cfdfcd60326c. Since then, various other macros have been improved to properly work with brace groups enabled or disabled.

We can now remove the brace group using variant of STMT_START/STMT_END, which will silence the warnings from clang.

demerphq commented:
this is failing CI, but it looks ok to me, not sure what is up.

atoomic commented on dist/Devel-PPPort/parts/inc/misc:
```diff
@@ -299,18 +299,8 @@ __UNDEF_NOT_PROVIDED__  PERL_GCC_BRACE_GROUPS_FORBIDDEN
-#  endif
-#endif
+#define STMT_START do
+#define STMT_END while (0)
```
Wasn't there any good reasons to prefer the other flavor when possible?

khwilliamson replied:
I thought there were multiple implementations depending on the compiler (besides the brace groups).  Perhaps those went away with C99?

haarg replied:
I've restored the other variant of this. I'm not certain of the exact necessity of it, but it seems reasonable to keep.

haarg commented:
After going over the history of this, I noted that this had previously been changed and then reverted. Since then, there have been more changes to make the brace groups less needed.

I've restored the alternate non-brace group implementation of `STMT_START`/`STMT_END`, and updated the commit message to reflect the history of this.

demerphq commented on dist/Devel-PPPort/parts/inc/call:
```diff
@@ -54,7 +54,7 @@ __UNDEFINED__ PERL_LOADMOD_IMPORT_OPS   0x4
      && (_errsv = ERRSV)                            \
      && (SvROK(_errsv) || SvTRUE(_errsv))           \
-     && (croak_sv(_errsv), 1));                     \
+     && (({ croak_sv(_errsv); }), 1));              \
```
This seems weird to me, wouldnt it be simpler to just define a croak_sv_ret() that returns 1? It would never actually do that as it would longjump, but wouldn't it be a simpler way to avoid the need for the list or the brace groups? It strikes me as something that is likely useful in other marcos.

haarg replied:
In core, `croak_sv` is a function. Devel::PPPort provides a version of it, but it's just a macro using `STMT_START`/`STMT_END`. That feels wrong. Maybe it could be changed like so:
```
PERL_STATIC_INLINE void D_PPP_CROAK_SV(pTHX_ SV *sv) {
    SV *_sv = (sv);
    if (SvROK(_sv)) {
        sv_setsv(ERRSV, _sv);
        croak(NULL);
    } else {
        D_PPP_FIX_UTF8_ERRSV_FOR_SV(_sv);
        croak("%" SVf, SVfARG(_sv));
    }
}
#  define croak_sv(sv) D_PPP_CROAK_SV(sv)
```

demerphq replied:
My point is that I think the (croak_sv(),1) is there because Perl_croak_sv() does not return a value. It is not a true function, but rather a procedure with a void return. So you cant use it in an expression by itself like the one we are looking at. If we create a Perl_croak_sv_ret() which returns 0, or 1, or whatever (it doesnt matter it will never return). Then the compile will be happy and there would be no need for (croak_sv(),1).

Eg, if a function is void then i think 

```
int x = void_function();
```

is an error but

```
int x = (void_function(),1);
```
is not.

I think that the function you propose would have the same problem, even if it works around the issue with PPPort providing a macro form of it.

demerphq replied:
Frankly it seems weird to me at a certain level that the croak() functions dont return something. It make them usable in expressions without hijinks, and the return can be casted away where we dont need it. I guess that boat has long sailed now tho.

haarg committed:
Devel-PPPort: use inline function for croak_sv without brace groups

The macro for croak_sv when brace groups are not enabled uses
STMT_START/STMT_END. This works when used as a statement, but would
break if used as part of an expression.

To resolve this, change that macro to use a function. As a static inline
function with a namespaced name, it shouldn't cause problems with
namespace pollution.

haarg committed:
Devel-PPPort: fix STMT_START and STMT_END to not warn on clang

Since 7169efc77525df70484a824bff4ceebd1fafc760, perl's core STMT_START
and STMT_END macros no longer try to use brace groups, due to the
warnings they can generate and their very limited usefulness.

That commit also changed Devel::PPPort to remove the use of brace groups
in STMT_START/STMT_END. That led to errors in older perls, so it was
partly reverted in e08ee3cb66f362c4901846a46014cfdfcd60326c. Since then,
various other macros have been improved to properly work with brace
groups enabled or disabled.

We can now remove the brace group using variant of STMT_START/STMT_END,
which will silence the warnings from clang.

haarg committed:
Devel-PPPort: bump version

haarg commented:
I changed this to define `croak_sv` as a macro to a static inline function. Checking over the history, Devel::PPPort has been changed to provide several other functions in the same way, so it seems it should be safe.

This allows removing the brace groups from STMT_START/STMT_END without any other problems.
